google-ads-examples/src/main/java/com/google/ads/googleads/examples/accountmanagement/GetAccountHierarchy.java

```diff
@@ -0,0 +1,341 @@
+              + "accounts or accounts with setup issues. Please check the logs for details.");
+      for (long accountId : accountsWithNoInfo) {
+        System.out.println(accountId);
+      }
+      System.out.println();
+    }
+
+    int depth = 0;
+    // Prints the hierarchy information for all accounts for which there is hierarchy information
+    // available.
+    for (CustomerClient rootCustomerClient : allHierarchies.keySet()) {
+      System.out.printf("Hierarchy of customer ID %d:%n", rootCustomerClient.getId().getValue());
+      printAccountHierarchy(rootCustomerClient, allHierarchies.get(rootCustomerClient), depth);
+      System.out.println();
+    }
+  }
+
+  /**
+   * Creates a map between a CustomerClient and each of its managers' mappings.
+   *
+   * @param loginCustomerId the loginCustomerId used to create the GoogleAdsClient.
+   * @param seedCustomerId the ID of the customer at the root of the tree.
+   * @return a map between a CustomerClient and each of its managers' mappings if the account
+   *     hierarchy can be retrieved. If the account hierarchy cannot be retrieved, returns null.
+   * @throws IOException if a Google Ads Client is not successfully created.
+   */
+  private Map<CustomerClient, Multimap<Long, CustomerClient>> createCustomerClientToHierarchy(
+      Long loginCustomerId, long seedCustomerId) throws IOException {
+    Set<Long> managerAccountsToSearch = new HashSet<>();
+    CustomerClient rootCustomerClient = null;
+
+    // Creates a GoogleAdsClient with the specified loginCustomerId. See
+    // https://developers.google.com/google-ads/api/docs/concepts/call-structure#cid for more
+    // information.
+    GoogleAdsClient googleAdsClient =
+        GoogleAdsClient.newBuilder()
+            .fromPropertiesFile()
+            .setLoginCustomerId(loginCustomerId != null ? loginCustomerId : seedCustomerId)
+            .build();
+
+    // Creates the Google Ads Service client.
+    try (GoogleAdsServiceClient googleAdsServiceClient =
+        googleAdsClient.getLatestVersion().createGoogleAdsServiceClient()) {
+      // Creates a query that retrieves all child accounts of the manager specified in search
+      // calls below.
+      String query =
+          "SELECT customer_client.client_customer, customer_client.level, "
+              + "customer_client.manager, customer_client.descriptive_name, "
+              + "customer_client.currency_code, customer_client.time_zone, "
+              + "customer_client.id "
+              + "FROM customer_client "
+              + "WHERE customer_client.level <= 1";
+
+      // Adds the seed customer ID to the list of IDs to be processed.
+      managerAccountsToSearch.add(seedCustomerId);
+      // Performs a breadth-first search algorithm to build a mapping of managers to their
+      // child accounts.
+      Multimap<Long, CustomerClient> customerIdsToChildAccounts = ArrayListMultimap.create();
+      while (!managerAccountsToSearch.isEmpty()) {
```

Comment:
Here I think the logic gets a little confusing - it looks like a clearer representation here would be:

`Queue<Long> managerIdsToSearch = new ArrayBlockingQueue();`

Then pulling from the top of the queue is just `managerIdsToSearch.poll()`.

Reply:
I think the logic is pretty straight-forward here. Also you must initialize ArrayBlockingQueue with a capacity, which we don't know at the outset because we don't know how many accounts in a given hierarchy are managers.

Reply:
The thing I find a bit confusing here is that we're checking if the queue is empty, creating and iterator and throwing it away immediately. It looks like the Queue interface is actually what we're implementing here, not a Set.

Also sorry I just rhymed off ArrayBlockingQueue on autopilot - I think LinkedList() would be the right choice :-)